altmetric/explorer/api/filters.py

```diff
@@ -0,0 +1,63 @@
+class Filters:
+    '''Holds a list of filters and provides methods that transform them
+     into query parameters and message digests.'''
+
+    def __init__(self):
+        """Initializes a new Filters object"""
+        self.items = []
+
+    def add_filter(self, arg, value):
+        """Add a new filter to the list of filters
+
+        Args:
+            arg (string): name of the filter
+            value (string, list, tuple or set): filter value(s)
+
+        Returns:
+            Filters: self
+        """
+        self.items.append((arg, value))
+        return self
+
+    def __len__(self):
+        """Count of filters
+
+        Returns:
+            integer: number of filters stored
+        """
+        return len(self.items)
+
+    def message(self):
+        """Convert the Filters to a message that can be used to construct a
+        digest for authentication.
+
+        See https://www.altmetric.com/explorer/documentation/api#authentication
+
+        Returns:
+            string: the digest
+        """
+        result = []
+        for arg, value in sorted(self.items):
+            result.append(arg)
+            if type(value) in (list, tuple, set):
+                for val in value:
+                    result.append(val)
+            else:
+                result.append(value)
+        return '|'.join(result)
+
+    def __str__(self):
+        """Create a string representation of the Filters in the form of query
+        parameters for convenience.
+
+        Returns:
+            string: api query parameters
+        """
+        result = []
+        for arg, value in self.items:
```

Comment:
Interesting!  I originally wrote that the example would be hard to make into a comprehension without introducing a separate method but then I noticed the duplication of ideas between lines 42 and 57 and that we actually have a bug because line 57 does not consider `set`s.

I've made a small change to extract the test for multi-item values into a separate function that both these methods can use but (as we discussed earlier) I can't see an easy way to convert the logic to a comprehension without losing readability. Feel free to have a go though!

NB `reduce` [does exist ](https://realpython.com/python-reduce-function/) but it's hidden away in a library and it's not as flexible as it is in ruby or JS because pythons anonymous functions ([see the `lambda` keyword](https://realpython.com/python-lambda/)) are limited to one line expressions and cannot contain statements like `if` so you can only do very simple things with them.  The alternative is to pass a 'real' function to the lambda but then you're back to the same problem as with a comprehension where you're adding extra function(s) and losing readability just so you can use `reduce`.